### PR TITLE
20161223 PR

### PR DESCRIPTION
**Pull request by hans-maws**
--> Added a **validateSignatureForGet** and a **validateSignatureForPost** method into AuthyUtil component to help the user validate the signature of the callback for OneTouch without too much problems or need to understand encryption.
--> api_key and api_url are now loaded from environment variables or properties file with **loadProperties** method from **AuthyUtil** (*environment variables will override properties file*)
--> Fixed a bad issue were parameters were not being properly set by the Resource.prepareGet method so options like force=true were never sent.
--> Updated README.md file with one touch information.
--> Added some extra Unit testing cases./
--> Fixed from String comparison issues (using == instead of .equals for Strings)
--> Did an initial code refactor.
--> Updated existent JUnit test cases that were failing with the updated messages.

### CODE REVIEW

File: src/test/resources/com/authy/api/test.properties [hunk 1]
```diff
@@ -1,7 +1,9 @@
 #place your api key here, get it from: https://dashboard.authy.com/ and go to your application dashboard
-api_key=xzfVSc8CTwU2Mpl959CcMhw3fZBfI5F0
+api_key=yourkeygoeshere
 api_url=https://api.authy.com/
 #place the id of the user you want to use for testing.
-user_id=2888888
-onetouch_udid=abcf88ab0-a8af-0134-087f-0ae83edadd62
+user_id=28894864
+#this will be just an UDID user for the unit test, it will be not valid for your app.
+onetouch_udid=ee3aff60-a8f9-0134-aed0-0a5fb1fd6492
```
Comment (vargasx) [inline src/test/resources/com/authy/api/test.properties]: please find all udid typos ?

File: src/test/java/com/authy/api/TestPhoneVerification.java [hunk 2]
```diff
@@ -44,18 +33,18 @@ public void itTestsVerificationStartEs() {
...
 
     @Test
     public void itTestsVerificationStartEn() {
         Params params = new Params();
         params.setAttribute("locale", "en");
-        Verification result = client.start("555-555-5555", "1", "sms", params);
+        Verification result = client.start("+1 775-461-5609", "1", "sms", params);
 
-        Assert.assertEquals("Text message sent to +1 555-555-5555.", result.getMessage());
+        Assert.assertEquals("Text message sent to +1 775-461-5609.", result.getMessage());
```
Comment (vargasx) [inline src/test/java/com/authy/api/TestPhoneVerification.java]: can we move country codes and phone numbers to properties as well ?

what do you think?

File: src/main/java/com/authy/AuthyUtil.java [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+     * Validates the request information to
+     *
+     * @param body       The body of the request in case of a POST method
+     * @param headers    The headers of the request
+     * @param url        The url of the request.
+     * @param authyToken the security token from the authy library
+     * @return true if the signature ios valid, false otherwise
+     * @throws UnsupportedEncodingException if the string parameters have problems with UTF-8 encoding.
+     */
+    public static boolean validateSignatureForPost(String body, Map<String, String> headers, String url, String authyToken) throws UnsupportedEncodingException {
```
Comment (vargasx) [inline src/main/java/com/authy/AuthyUtil.java]: could you add unit test for this ?

File: src/main/java/com/authy/AuthyUtil.java [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+     * Validates the request information to
+     *
+     * @param params     The query parameters in case of a GET request
+     * @param headers    The headers of the request
+     * @param url        The url of the request.
+     * @param authyToken the security token from the authy library
+     * @return true if the signature ios valid, false otherwise
+     * @throws UnsupportedEncodingException if the string parameters have problems with UTF-8 encoding.
+     */
+    public static boolean validateSignatureForGet(Map<String, String> params, Map<String, String> headers, String url, String authyToken) throws UnsupportedEncodingException {
```
Comment (vargasx) [inline src/main/java/com/authy/AuthyUtil.java]: could you add unit test for this ?

File: README.md [hunk 5]
```diff
@@ -179,6 +179,80 @@ If the request was successful, you will need to store the authy id in your datab
...
 ```
 
+## OneTouch Support
+
+  `client.getOneTouch()` will return the OneTouch wrapper, with this component you will be able to use all the 
+  endpoints defined in the [OneTouch API](https://www.twilio.com/docs/api/authy/authy-onetouch-api), 
+  for example, using the 
+  'sendApprovalRequest' method you will create an ApprovalRequest
+  in OneTouch, then the approval request
+  status can be queried by using the  `uuid` provided. (NOTE: The UDID is the unique identifier for the request sent to OneTouch)
```
Comment (vargasx) [inline README.md]: udid typo?

File: README.md [hunk 6]
```diff
@@ -179,6 +179,80 @@ If the request was successful, you will need to store the authy id in your datab
...
+    OneTouchResponse response = client.getOneTouch().sendApprovalRequest(Integer.parseInt(103), "Authorize OneTouch Unit Test", details, hidden, logos, 86400);
+    // If the request was successfuly created.
+    assert response.isSuccess();
+    String uuid = response.getApprovalRequest().getUUID();
+    System.out.println(uuid);
+```
+
+  If you want to query the ApprovalRequest status
+```java
+  String status = client.getOneTouch().getApprovalRequestStatus(udid).getStatus();
```
Comment (vargasx) [inline README.md]: udid typo?

File: README.md [hunk 7]
```diff
@@ -179,6 +179,80 @@ If the request was successful, you will need to store the authy id in your datab
...
+  System.out.println(status);
+```
+When you make a OneTouch request, you will get a OneTouchResponse object which in turn will have a getApprovalRequest() accessor to retrieve the metadata from the given request, for example:
+
+```java
+    OneTouchResponse response = client.getOneTouch().getApprovalRequestStatus(udid);
```
Comment (vargasx) [inline README.md]: udid typo?

File: README.md [hunk 8]
```diff
@@ -179,6 +179,80 @@ If the request was successful, you will need to store the authy id in your datab
...
+    ApprovalRequest approvalRequest  = response.getApprovalRequest();
+    // was the user notified?
+    System.out.println("The user was "+(approvalRequest.isNotified()?"":"not")+" notified.");
+    // What is the status of the request?
+    System.iout.print("The status of the request is: "+approvalRequest.getStatus());
```
Comment (vargasx) [inline README.md]: .iout typo?

File: src/test/resources/com/authy/api/test.properties (hunk 1 above)
Comment (hans-maws) [reply src/test/resources/com/authy/api/test.properties]: will do

File: README.md (hunk 5 above)
Comment (hans-maws) [reply README.md]: on it

File: README.md (hunk 6 above)
Comment (hans-maws) [reply README.md]: on it

File: README.md (hunk 7 above)
Comment (hans-maws) [reply README.md]: on it

File: README.md (hunk 8 above)
Comment (hans-maws) [reply README.md]: done

File: src/test/java/com/authy/api/TestPhoneVerification.java (hunk 2 above)
Comment (hans-maws) [reply src/test/java/com/authy/api/TestPhoneVerification.java]: done

File: README.md [hunk 9]
```diff
@@ -179,6 +179,80 @@ If the request was successful, you will need to store the authy id in your datab
...
+    details.put("username", "Bill Smith");
+    details.put("location", "California, USA");
+    details.put("amount", "$20.000");
+
+    HashMap<String, String> hidden = new HashMap<String, String>();
+    hidden.put("ip_address", "10.10.3.203");
+
+    HashMap<String, String> logos = new HashMap<String, String>();
+
+    OneTouchResponse response = client.getOneTouch().sendApprovalRequest(Integer.parseInt(103), "Authorize OneTouch Unit Test", details, hidden, logos, 86400);
```
Comment (vargasx) [inline README.md]: Could you please see this method signature from ruby `https://github.com/authy/authy-ruby/blob/master/lib/authy/onetouch.rb#L12` it only receives a hash. could we change it here to receive one hash and validate it contains authy_id, message, also could you check how we validate logos in Ruby lib in same file `https://github.com/authy/authy-ruby/blob/master/lib/authy/onetouch.rb`

File: README.md [hunk 10]
```diff
@@ -179,6 +179,80 @@ If the request was successful, you will need to store the authy id in your datab
...
+    // If the request was successfuly created.
+    assert response.isSuccess();
+    String uuid = response.getApprovalRequest().getUUID();
```
Comment (vargasx) [inline README.md]: I'm trying to use this code in eclipse but I'm getting this message `The type OneTouchResponse.ApprovalRequest is no visible`

File: README.md (hunk 9 above)
Comment (vargasx) [reply README.md]: @hans-maws lets wait before you make changes here base on suggestion. I'll review and make a further comment.

**Pull request merged**